src/coreclr/jit/flowgraph.cpp

```diff
@@ -436,6 +436,228 @@ BasicBlock* Compiler::fgCreateGCPoll(GCPollType pollType, BasicBlock* block)
+            {
+                bool isGc = false;
+                if (!IsStaticHelperEligibleForExpansion(tree, &isGc))
+                {
+                    continue;
+                }
+
+                GenTreeCall* call = tree->AsCall();
+                if (call->IsTailCall())
+                {
+                    // It is very unlikely, but just in case
+                    continue;
+                }
+
+                assert(call->gtRetClsHnd != NO_CLASS_HANDLE);
+                JITDUMP("Expanding static initialization for %s\n", eeGetClassName(call->gtRetClsHnd))
+
+                UINT8          isInitMask;
+                InfoAccessType staticBaseAccessType;
+                size_t         staticBase;
+                size_t         staticInitAddr =
+                    info.compCompHnd->getIsClassInitedFieldAddress(call->gtRetClsHnd, isGc, &staticBaseAccessType,
+                                                                   &staticBase, &isInitMask);
+                if (staticInitAddr == 0)
+                {
+                    continue;
+                }
+                assert((staticBase != 0) && (isInitMask != 0));
+
+                DebugInfo debugInfo = stmt->GetDebugInfo();
+
+                // Split block right before the call tree
+                BasicBlock* prevBb       = block;
+                GenTree**   callUse      = nullptr;
+                Statement*  newFirstStmt = nullptr;
+                block                    = fgSplitBlockBeforeTree(block, stmt, call, &newFirstStmt, &callUse);
+                assert(prevBb != nullptr && block != nullptr);
+
+                // Block ops inserted by the split need to be morphed here since we are after morph.
+                // We cannot morph stmt yet as we may modify it further below, and the morphing
+                // could invalidate callUse.
+                while ((newFirstStmt != nullptr) && (newFirstStmt != stmt))
+                {
+                    fgMorphStmtBlockOps(block, newFirstStmt);
+                    newFirstStmt = newFirstStmt->GetNextStmt();
+                }
+
+                // Define a local for the result
+                const unsigned resultNum   = lvaGrabTemp(true DEBUGARG("static init"));
+                lvaTable[resultNum].lvType = call->TypeGet();
+                GenTreeLclVar* resultLcl   = gtNewLclvNode(resultNum, call->TypeGet());
+
+                // Replace current use with a temp local. That local will be set in either
+                // fastPathBb or fallbackBb
+                *callUse = gtClone(resultLcl);
+
+                fgMorphStmtBlockOps(block, stmt);
+                gtUpdateStmtSideEffects(stmt);
+
+                //
+                // Create new blocks. Essentially, we want to transform this:
+                //
+                //     result = helperCall();
+                //
+                // into:
+                //
+                //     if (isInited)
```

Comment:
This should generate smaller code. Would this be better?
   ```
                //     if (isInited)
                //     {
                //         goto Done;
                //     }
                //     helperCall();
                // Done:
                //     result = fastPath;
```

For throughput, it would be even better to do this:
                //     if (!isInited)
                //     {
                //         goto Initialize;
                //     }
                // Done:
                //     
...
                // rarely executed part of the method
                //     helperCall();
                //     goto Done:
```

Reply:
@jkotas the shape you proposed can be achieved naturally if I do `fallbackBb->bbSetRunRarely();` instead of applying 20% of weight from the source block. I was keeping in mind your suggestion from the previous PR that such thing might regress startup since the first execution will be longer

Reply:
> I was keeping in mind your suggestion from the previous PR

What was the previous PR?

Reply:
> > I was keeping in mind your suggestion from the previous PR
> 
> What was the previous PR?

https://github.com/dotnet/runtime/pull/81635#issuecomment-1445475674

> Rarely executed block (this case): It should be moved out-of-line for throughput optimizations, but it should stay in the **non-cold section for AOT workingset/startup optimizations**

Reply:
It also comes with a small size increase due to extra "jump back", e.g. currently for:
```csharp
    static int field = int.Parse("42"); // don't let NAOT to preinit it

    [MethodImpl(MethodImplOptions.NoInlining)]
    int Test()
    {
        return field;
    }
```
for JIT I have:
```asm
; Method Prog:Test():int:this
G_M3272_IG01:
       sub      rsp, 40
G_M3272_IG02:
       test     byte  ptr [(reloc)], 1      ; const ptr
       jne      SHORT G_M3272_IG04
G_M3272_IG03:
       mov      rcx, 0xD1FFAB1E
       mov      edx, 3
       call     CORINFO_HELP_GETSHARED_NONGCSTATIC_BASE
G_M3272_IG04:
       mov      eax, dword ptr [(reloc)]      ; static handle
G_M3272_IG05:
       add      rsp, 40
       ret      
; Total bytes of code: 44
```
When I mark the block as rarely-taken:
```asm
; Method Prog:Test():int:this
G_M3272_IG01:
       sub      rsp, 40
G_M3272_IG02:
       test     byte  ptr [(reloc)], 1      ; const ptr
       je       SHORT G_M3272_IG05
G_M3272_IG03:
       mov      eax, dword ptr [(reloc)]      ; static handle
G_M3272_IG04:
       add      rsp, 40
       ret      
G_M3272_IG05:
       mov      rcx, 0xD1FFAB1E
       mov      edx, 3
       call     CORINFO_HELP_GETSHARED_NONGCSTATIC_BASE
       jmp      SHORT G_M3272_IG03
; Total bytes of code: 46
```

Reply:
> non-cold section for AOT startup workingset optimizations

We do not have full hot/cold splitting enabled currently. I would suggest that we introduce a new `bbSetRunTypicallyOnce` method and call it for situation like this. For now, this method can just call `bbSetRunRarely` assuming that it always does the right thing as a throughput optimization.

If/once we start to care about AOT startup workingset optimizations, we can make the `bbSetRunTypicallyOnce` method smarter and do something different.

I think it is worth it to pay the few extra bytes of code for the better layout.